File: js/Practice2/DynamicProgramming/Other/PlayingCardsInsertionSort.js
// https://atcoder.jp/contests/abc006/tasks/abc006_4

Object.prototype.let = function(f) {
  return f(this.valueOf());
}

require('fs')
  .readFileSync('/dev/stdin', 'utf8')
  .trim()
  .split('\n')
  .map(x => +x)
  .let(arr => [
    arr[0],
    arr.slice(1),
    [...Array(arr[0] + 1)].map(_ => -1),
  ])
  .let(([n, ns, memo]) =>
    ns.reduce((len, x) =>
      memo[len] < x
        ? (_ => len + 1)(
          memo[len + 1] = x
        )
        : (_ => len)(
          memo.findIndex(x2 => x <= x2)
            .let(i => memo[i] = x)
        )
      , 0
    )
    .let(len => n - len)
  )
  .let(console.log);

// https://atcoder.jp/contests/abc006/submissions/21493390

File: fs/Practice2/DynamicProgramming/Other/PlayingCardsInsertionSort.fs
// https://atcoder.jp/contests/abc006/tasks/abc006_4

stdin.ReadLine ()
|> int
|> fun n ->
  (
    List.init n (fun _ -> stdin.ReadLine () |> int),
    Array.create (n + 1) -1
  )
  |> fun (ns, memo) ->
    ns
    |> List.fold (
      fun len x ->
        if memo.[len] < x
        then
          (memo.[len + 1] <- x)
          |> fun _ -> len + 1
        else
          memo
          |> Array.findIndex ((<=) x)
          |> fun i -> memo.[i] <- x
          |> fun _ -> len
    ) 0
  |> fun len -> n - len
|> printfn "%d"

// https://atcoder.jp/contests/abc006/submissions/21492908
